### PR TITLE
Add framework for running MSQ tests with taskSpec instead of SQL

### PR DESCRIPTION
**Pull request by adarshsanjeev**
This PR makes two changes:
 - Adds a function to fetch DataSegment to DataSegmentProvider, this allows MSQ tasks to fetch DataSegment.
 - Makes changes to the test framework to allow unit tests to be run against task specs as well. This will allow direct testing of queries which do not have a valid sql.

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by kfaraz** [comment]
@adarshsanjeev , why do we want to test MSQ specs that do not have a valid SQL?
Aren't all uses of the MSQ engine meant to be via SQL?

**Comment by cryptoe** [comment]
@kfaraz Wont this be required for MSQ autocompaction ?

**Comment by kfaraz** [comment]
Yes, @cryptoe, you are right. I was under the impression that we submit MSQ SQL for compaction but just checked with @gargvishesh and he confirmed that we submit a spec, not SQL.

Sorry for the confusion 🙇🏻

File: extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/DataSegmentProvider.java [hunk 1]
```diff
@@ -40,4 +41,9 @@ Supplier<ResourceHolder<Segment>> fetchSegment(
       ChannelCounters channelCounters,
       boolean isReindex
   );
+
+  DataSegment fetchDataSegment(
```
Comment (cryptoe) [inline extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/DataSegmentProvider.java:45]: I donot think these changes should be bundled with this PR.

Comment (adarshsanjeev) [reply extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/DataSegmentProvider.java]: Moved these changes out. Will raise a separate PR.

**Review by cryptoe — APPROVED** [review]

**Pull request merged**